### ISSUE TITLE
Enhancement Request: Custom error function for `lua_pcall`.

### ISSUE THREAD

**Issue by rpatters1** [kunitoki/LuaBridge3#158]
I have a custom error handler that I use for the main execution flow of my scripts (when I call `lua_pcall` at the top level). Its main purpose is to capture the line number and source code block where an error occurs:

```c++
int LuaErrorHandler(lua_State *L)
{
   MyScriptData* pScript = getActiveScript(L);
   if (pScript)
   {
      lua_Debug ar;
      int level = 0;
      while (lua_getstack(L, ++level, &ar))
      {
         lua_getinfo(L, "Sl", &ar);  /* get info about it */
         std::string source = ar.source;
         if (!source.empty() && source[0] == '@')
            source.erase(0, 1); // Erase the character at the front

         bool useThis = (ar.currentline > 0);
         if (useThis)
         {
            if (pScript->m_scriptText.has_value())
                useThis = (source == pScript->m_scriptText.value());
            else if (!pScript->m_finaleplugin.m_useString)
                useThis = (source == pScript->m_fullPath);
         }
         if (useThis)
         {
            pScript->m_lineNumber = ar.currentline;
            pScript->m_sourceName = ar.source;
            break;
         }
      }
   }
   return 1;
}


lua_pushcfunction(L, LuaErrorHandler);
int customHandlerIndex = lua_gettop(L); // Get the index of the custom error handler
int retval = luaL_loadfile(L, myscriptfile);
if (! retval)
   return lua_pcall(L, 0, LUA_MULTRET, customHandlerIndex); // launches my script
// do some other error handling here
```

The problem I'm having is that there is no way to assign this function when I do a callback with `LuaRef::operator()`. It would be nice if there were an option somewhere to do this. Ideally (at least for me), it would be done once when setting up the Lua state, and then it would be used on every call to `lua_pcall` in LuaBridge.

**Comment by kunitoki** [comment]
Yeah this is something that i always replace (have my own version of the LuaRef::operator()) i can roll something but it might be a dedicated method instead of adding more overload of the operaotor.

**Comment by rpatters1** [comment]
I was imagining registering the function some way and having it automatically be picked up by the operator(). I realize this may be less straightforward than it sounds.

**Comment by kunitoki** [comment]
I was thinking to have a `luabridge::invoke` and `luabridge::safeInvoke` or maybe i do a special tag dispatched operator() in LuaRef...

**Comment by rpatters1** [comment]
I suppose I'd have to see an example of what you are talking about to understand the proposal, but any solution is preferable to none.

The reason I was hoping for a 1-time registration is because of my dependency tree. It will be a somewhat messy if I have to supply the function where I invoke the callback. Nothing impossible, but just rather inelegant.

**Comment by rpatters1** [comment]
Thank you for tackling it. I'm getting an unused parameter on `errorTag`, which is easy enough to fix. But I'm wondering if it isn't revealing a flaw in the design.

Perhaps instead of an overload of `operator()` it should be its own thing, e.g. (from your test example):

```
    auto luaHandler = luabridge::newFunction(L, handler);
    luaHandler.push();

    f.callWithHandler(-2, "badly");
```

This makes plain what's going on, rather than the awkward passing of a special constant.

I realize there is already an inline `callWithHandler` function, but I mean a class method (with any other name if you don't like that one). Is there any reason to have both inline `call` and `callWithHandler`? It looks like `call` could simply be `callWithHandler` with a 0 handler.

As for your TODO question, I don't think you should call the error handler when `lua_pcall` returns. At least in my use case it would be counterproductive. The whole reason I need the error handler is to be able to traverse the Lua stack when the error occurs to capture the relevant line number and source file for reporting purposes. I then let my error handler fall through and error out the normal way. When `lua_pcall` returns, the stack information is gone.
 
Whatever you decide, I am sure I can make this work with my code.

**Comment by rpatters1** [comment]
Another thing that would help my use case is if instead of a stack location, I could pass a lua_CFunction as the error handler. Then LB could manage the stack. Managing the stack involves a lot of smelly hardcoded numbers, and it would be nice to keep those as encapsulated as possible.

**Comment by rpatters1** [comment]
It looks excellent to me now. Thank you!

**Comment by rpatters1** [comment]
I am finally getting around to implementing this, and I realized I am going to have to do runtime linking, due to the dependency stack between my Lua interpreter and the framework it wraps.  So that means I have something like:

```c++
lua_CFunction errorHandler = nullptr;
.
.
.
luaref.callWithHandler(errorHandler, args);
```

This compiles perfectly, but it crashes the system if `errorHandler` has not been populated. I could work around it by:

```
if (errorHandler)
   luaref.callWithHandler(errorHandler, args);
else
   luadref(args);
```

But that introduces ugly if statements in a routine that already has several flavors of luaref invocations. How would you feel about adding a runtime check for null in `checkWithHandler`? Otherwise there needs to be a way for this not to compile.

**Comment by rpatters1** [comment]
Another issue I'm having is that if I supply the error function, `callWithHandler` does not throw the error. My "error handler" function is simply capturing information about the stack that is not available after the `lua_pcall` returns. I have not found any better way to do it, but this works quite well. However, that means I am still counting on the normal error exit with `throw`.

I suppose I can throw it myself after `callWithHandler` returns. Perhaps it isn't a big deal, but I thought I would mention it to get your opinion.
